sdk/core/core-rest-pipeline/documentation/pipeline.md

```diff
@@ -0,0 +1,220 @@
+
+export interface PipelinePolicy {
+  name: string;
+  sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse>;
+}
+```
+
+It is given a string name so that consumers can express dependencies relative to existing policies. The `sendRequest` method uses a middleware pattern, which is familiar to JS developers.
+
+# Pipelines
+
+A `Pipeline` composes several policies and provides the ability to make a request that will flow through the policies in a specific order determined by their requirements. Which allows manipulating each request before and after it is made to the server.
+
+```typescript
+export interface AddPipelineOptions {
+  afterPhase?: PipelinePhase;
+  afterPolicies?: string[];
+  beforePolicies?: string[];
+  phase?: PipelinePhase;
+}
+
+export interface Pipeline {
+  addPolicy(policy: PipelinePolicy, options?: AddPipelineOptions): void;
+  clone(): Pipeline;
+  getOrderedPolicies(): PipelinePolicy[];
+  removePolicy(options: { name?: string; phase?: PipelinePhase }): PipelinePolicy[];
+  sendRequest(httpClient: HttpClient, request: PipelineRequest): Promise<PipelineResponse>;
+}
+```
+
+Dependencies expressed when calling `addPolicy` are not sensitive to order. The following code snippets have identical effects:
+
+```typescript
+pipeline.addPolicy(barPolicy);
+pipeline.addPolicy(fooPolicy, { afterPolicy: "barPolicy" });
+
+// The below is the same as the above.
+// Policies don't have to be part of the pipeline in order to be
+// referenced by before/afterPolicy.
+pipeline.addPolicy(fooPolicy, { afterPolicy: "barPolicy" });
+pipeline.addPolicy(barPolicy);
+```
+
+Phases are predefined buckets of activity that pre-defined policies inside core will fall into. They are currently a fixed set and not customizable by pipeline consumers.
+
+```typescript
+export type PipelinePhase = "Deserialize" | "Serialize" | "Retry" | "Sign";
+```
+
+# Default Pipeline
+
+`@azure/core-rest-pipeline` exposes `createPipelineFromOptions` which creates pipeline with a predefined set of policies. The resulting pipeline can be configured through `PipelineOptions`
+
+```typescript
+/**
+ * Defines options that are used to configure the HTTP pipeline for
+ * an SDK client.
+ */
+export interface PipelineOptions {
+  /**
+   * Options that control how to retry failed requests.
+   */
+  retryOptions?: PipelineRetryOptions;
+
+  /**
+   * Options to configure a proxy for outgoing requests.
+   */
+  proxyOptions?: ProxySettings;
+
+  /**
+   * Options for how redirect responses are handled.
+   */
+  redirectOptions?: RedirectPolicyOptions;
+
+  /**
+   * Options for adding user agent details to outgoing requests.
+   */
+  userAgentOptions?: UserAgentPolicyOptions;
+}
+
+/**
+ * Defines options that are used to configure internal options of
+ * the HTTP pipeline for an SDK client.
+ */
+export interface InternalPipelineOptions extends PipelineOptions {
```

Comment:
I think some remarks about how we use this inside SDK clients but don't expose it to folks using them would be good